test/kubernetes/deployment.go

```diff
@@ -0,0 +1,222 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networking "k8s.io/api/networking/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+// EchoService name of the deployment for the echo app
+const EchoService = "echo"
+
+// EchoContainer container image name
+//const EchoContainer = "gcr.io/k8s-staging-ingressconformance/echoserver@sha256:e4e132da40d303f4b50e183b977ed1eb5c6db6eb0ddbdaa19565d14d49e05940"
+const EchoContainer = "aledbf/echoserver"
+
+const deploymentTemplate = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: %v
+spec:
+  replicas: 1
+  strategy:
+    type: RollingUpdate
+  selector:
+    matchLabels:
+      app: %v
+  template:
+    metadata:
+      labels:
+        app: %v
+    spec:
+      containers:
+      - name: ingress-conformance-echo
+        image: %v
+        env:
+        - name: POD_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.name
+        - name: NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        - name: INGRESS_NAME
+          value: %v
+        - name: SERVICE_NAME
+          value: %v
+        ports:
+        - name: %v
+          containerPort: 3000
+        readinessProbe:
+          httpGet:
+            path: /health
+            port: 3000
+`
+const serviceTemplate = `
+apiVersion: v1
+kind: Service
+metadata:
+  name: %v
+spec:
+  selector:
+    app: %v
+  ports:
+    - port: %v
+      targetPort: 3000
+`
+
+// NewEchoDeployment creates a new deployment of the echoserver image in a particular namespace.
+func NewEchoDeployment(kubeClientSet kubernetes.Interface, namespace, name, serviceName, servicePortName string, servicePort int32) error {
+	deploymentName := fmt.Sprintf("%v-%v", name, serviceName)
+	manifest := fmt.Sprintf(deploymentTemplate, deploymentName, deploymentName, deploymentName, EchoContainer, name, serviceName, servicePortName)
+	deployment, err := deploymentFromManifest(manifest)
+	if err != nil {
+		return err
+	}
+
+	_, err = applyDeployment(kubeClientSet, namespace, deployment)
+	if err != nil {
+		return fmt.Errorf("creating deployment (%v): %w", deployment.Name, err)
+	}
+
+	manifest = fmt.Sprintf(serviceTemplate, serviceName, deploymentName, servicePort)
+	service, err := serviceFromManifest(manifest)
+	if err != nil {
+		return err
+	}
+
+	if servicePortName != "" {
+		service.Spec.Ports[0].Name = servicePortName
+	}
+
+	// if no port is defined, use default 8080
+	if servicePort == 0 {
+		service.Spec.Ports[0].Port = 8080
+	}
+
+	service, err = applyService(kubeClientSet, namespace, service)
+	if err != nil {
+		return fmt.Errorf("creating service (%v): %w", service.Name, err)
+	}
+
+	err = waitForEndpoints(kubeClientSet, 5*time.Minute, service.Namespace, service.Name, 1)
```

Comment:
`5*time.Minute` should be extracted to a const.

Reply:
done. Extracted to a var and flag